crates/nri-resctrl-plugin/src/lib.rs

```diff
@@ -244,6 +466,61 @@ impl<P: FsProvider + Send + Sync + 'static> Plugin for ResctrlPlugin<P> {
         req: StateChangeEvent,
     ) -> ttrpc::Result<Empty> {
         debug!("resctrl-plugin: state_change: event={:?}", req.event);
+        match req.event.enum_value() {
+            Ok(Event::REMOVE_POD_SANDBOX) => {
+                if let Some(pod) = req.pod.as_ref() {
+                    let pod_uid = pod.uid.clone();
+                    let mut st = self.state.lock().unwrap();
+                    let group_path = st
+                        .pods
+                        .get(&pod_uid)
+                        .and_then(|ps| ps.group_path.clone());
+                    st.containers.retain(|_, c| c.pod_uid != pod_uid);
+                    st.pods.remove(&pod_uid);
+                    drop(st);
+                    if let Some(gp) = group_path {
+                        if let Err(e) = self.resctrl.delete_group(&gp) {
+                            warn!("resctrl-plugin: failed to delete group {}: {}", gp, e);
+                        }
+                        self.emit_event(PodResctrlEvent::Removed(PodResctrlRemoved {
+                            pod_uid,
+                            group_path: Some(gp),
+                        }));
+                    } else {
+                        self.emit_event(PodResctrlEvent::Removed(PodResctrlRemoved {
+                            pod_uid,
+                            group_path: None,
+                        }));
+                    }
+                }
+            }
+            Ok(Event::REMOVE_CONTAINER) => {
+                if let (Some(pod), Some(container)) = (req.pod.as_ref(), req.container.as_ref()) {
+                    let pod_uid = pod.uid.clone();
+                    let mut st = self.state.lock().unwrap();
+                    if let Some(cstate) = st.containers.remove(&container.id) {
+                        if let Some(ps) = st.pods.get_mut(&pod_uid) {
+                            if ps.total_containers > 0 {
+                                ps.total_containers -= 1;
+                            }
+                            if cstate.reconciled && ps.reconciled_containers > 0 {
+                                ps.reconciled_containers -= 1;
+                            }
+                            let ev = PodResctrlEvent::Added(PodResctrlAdded {
```

Comment:
1. let's call the message AddOrUpdate since we're using it for both
2. refactor sending the message for a given PodState into an internal function emit_pod_add_or_update(). Note that the state could also be failed. And let's replace all emissions of this message to use the helper function.
3. we need to emit the message also in handle_pod